### PR TITLE
Dynamic provisioner for GCE PD

### PR DESCRIPTION
**Pull request by pmorie**
Replaces #15604

@thockin @saad-ali @jsafrane @markturansky

### CODE REVIEW

File: pkg/volume/gce_pd/gce_util.go [hunk 1]
```diff
@@ -112,6 +112,40 @@ func (util *GCEDiskUtil) DetachDisk(c *gcePersistentDiskCleaner) error {
...
+func (gceutil *GCEDiskUtil) CreateVolume(c *gcePersistentDiskProvisioner) (volumeID string, volumeSizeGB int, err error) {
+	cloud, err := getCloudProvider()
+	if err != nil {
+		return "", 0, err
+	}
+
+	name := fmt.Sprintf("kube-dynamic-%s", util.NewUUID())
+	requestBytes := c.options.Capacity.Value()
+	// GCE works with gigabytes, convert to GB with rounding up
+	const giga = 1024 * 1024 * 1024
```
Comment (saad-ali) [inline pkg/volume/gce_pd/gce_util.go]: Should this be GiB or GB? Looks like GCE expects GB, if so is this conversion valid? Also, same comments as AWS and Cinder: seems like this should be common code, not custom for each cloud provider (standard conversion from bytes -> GB or GiB or MB or MiB etc).

File: pkg/volume/gce_pd/gce_pd.go [hunk 2]
```diff
@@ -301,3 +343,66 @@ func (c *gcePersistentDiskCleaner) TearDownAt(dir string) error {
...
+var _ volume.Provisioner = &gcePersistentDiskProvisioner{}
+
+func (c *gcePersistentDiskProvisioner) Provision(pv *api.PersistentVolume) error {
+	volumeID, sizeGB, err := c.manager.CreateVolume(c)
+	if err != nil {
+		return err
+	}
+	pv.Spec.PersistentVolumeSource.GCEPersistentDisk.PDName = volumeID
+	pv.Spec.Capacity = api.ResourceList{
+		api.ResourceName(api.ResourceStorage): resource.MustParse(fmt.Sprintf("%dGi", sizeGB)),
```
Comment (saad-ali) [inline pkg/volume/gce_pd/gce_pd.go]: I think GCE API return GB not GiB

File: pkg/volume/gce_pd/gce_pd.go [hunk 3]
```diff
@@ -122,12 +126,50 @@ func (plugin *gcePersistentDiskPlugin) newCleanerInternal(volName string, podUID
 	}}, nil
 }
 
+func (plugin *gcePersistentDiskPlugin) NewDeleter(spec *volume.Spec) (volume.Deleter, error) {
+	return plugin.newDeleterInternal(spec, &GCEDiskUtil{})
+}
+
+func (plugin *gcePersistentDiskPlugin) newDeleterInternal(spec *volume.Spec, manager pdManager) (volume.Deleter, error) {
```
Comment (saad-ali) [inline pkg/volume/gce_pd/gce_pd.go]: Why is the helper function needed? Why not just stick this logic in `NewDeleter`? Ditto with `NewProvisioner`.

**Comment by saad-ali** [comment]
Same comments across all three PRs about clarification/factoring of bytes -> GB/GiB/MB/MiB code. Other than that, looks good.

File: pkg/volume/gce_pd/gce_pd.go (hunk 2 above)
Comment (jsafrane) [inline pkg/volume/gce_pd/gce_pd.go]: No, it does not. All disk sizes are in GiB, just the documentation (and GUI and basically everything) says it's GB.

**Comment by jsafrane** [comment]
Build failure is expected, this PR depends on PR #18601.

File: pkg/volume/gce_pd/gce_pd.go (hunk 3 above)
Comment (jsafrane) [inline pkg/volume/gce_pd/gce_pd.go]: It's used in testing to inject fake pdManager.

Comment (saad-ali) [inline pkg/volume/gce_pd/gce_pd.go]: Ack.

File: pkg/volume/gce_pd/gce_pd.go (hunk 2 above)
Comment (saad-ali) [inline pkg/volume/gce_pd/gce_pd.go]: Nice

**Comment by saad-ali** [comment]
LGTM. I will add LGTM label after #18607 is merged and this PR is rebased/squashed.
Thanks for your patience @jsafrane!

**Comment by pmorie** [comment]
Applying auto-lgtm per discussion w/ @saad-ali

**Comment by pmorie** [comment]
@k8s-bot test this

**Comment by pmorie** [comment]
Ugh, gofmt

**Comment by pmorie** [comment]
@jsafrane I get a failure in the unit tests:

```
Running tests for APIVersion: v1,extensions/v1beta1,metrics/v1alpha1 with etcdPrefix: registry
+++ [1216 17:29:02] Running tests without code coverage
--- FAIL: TestPlugin (0.00s)
        gce_pd_test.go:235: Provision() returned unexpected volume size: 107374182400
FAIL
FAIL    k8s.io/kubernetes/pkg/volume/gce_pd     0.007s
!!! Error in hack/test-go.sh:240
```

**Comment by jsafrane** [comment]
> --- FAIL: TestPlugin (0.00s)
        gce_pd_test.go:235: Provision() returned unexpected volume size: 107374182400

Fixed. Thanks for the rebases @pmorie!

**Comment by jsafrane** [comment]
@pmorie @saad-ali: all green, ready to merge

**Pull request merged**